File: backend/app/voice/protocol.py
from __future__ import annotations

import math
import struct
from dataclasses import dataclass
from typing import Any

from app.voice.feedback import CUE_LANGUAGES

CAPTURE_MAGIC = b"ADSH"
CAPTURE_VERSION = 1
CAPTURE_HEADER = struct.Struct("!4sBQd")
CAPTURE_HEADER_BYTES = CAPTURE_HEADER.size


class ProtocolError(ValueError):
    pass


@dataclass(frozen=True)
class CaptureFrame:
    capture_seq: int
    capture_time_ms: float
    pcm: bytes


@dataclass(frozen=True)
class ProbeOptions:
    enabled: bool = False
    stt_input_audio_codec: str | None = None
    stt_audio_encoding: str | None = None
    tts_output_audio_codec: str | None = None
    endpointing_strategy: str | None = None


@dataclass(frozen=True)
class ClientToolResult:
    call_id: str
    result: Any


CLIENT_MESSAGE_TYPES = frozenset(
    {
        "end_call",
        "interrupt_ack",
        "playback_finished",
        "playback_started",
        "probe_flush",
        "response_cue_started",
        "tool_result",
    }
)

SERVER_MESSAGE_TYPES = frozenset(
    {
        "agent_state",
        "agent_text",
        "audio_start",
        "billing",
        "call_ended",
        "error",
        "final_transcript",
        "interrupt",
        "interrupt_resolved",
        "ready",
        "response_cue",
        "response_cue_cancel",
        "tool_request",
        "turn_complete",
        "turn_metrics",
    }
)

_CLIENT_REQUIRED_FIELDS: dict[str, dict[str, type | tuple[type, ...]]] = {
    "end_call": {},
    "interrupt_ack": {
        "turn_id": str,
        "interrupt_received_perf_ms": (int, float),
        "queue_cleared_perf_ms": (int, float),
        "audio_queue_cleared": bool,
    },
    "playback_finished": {
        "turn_id": str,
        "playback_end_perf_ms": (int, float),
        "played_audio_ms": (int, float),
    },
    "playback_started": {
        "turn_id": str,
        "last_speech_capture_seq": int,
        "audio_received_perf_ms": (int, float),
        "scheduled_perf_ms": (int, float),
        "playback_start_perf_ms": (int, float),
        "e2e_voice_to_voice_ms": (int, float),
    },
    "probe_flush": {},
    "response_cue_started": {
        "turn_id": str,
        "cue_id": str,
        "cue_start_perf_ms": (int, float),
        "feedback_voice_to_voice_ms": (int, float),
    },
    "tool_result": {"call_id": str, "result": object},
}

_SERVER_REQUIRED_FIELDS: dict[str, dict[str, type | tuple[type, ...]]] = {
    "agent_state": {"state": str},
    "agent_text": {"text": str},
    "audio_start": {
        "turn_id": str,
        "last_speech_capture_seq": int,
        "last_speech_capture_time_ms": (int, float),
    },
    "billing": {
        "seconds": int,
        "cost_paise": int,
        "balance_paise": int,
    },
    "call_ended": {"reason": str},
    "error": {"code": str, "message": str},
    "final_transcript": {"turn_id": str, "text": str},
    "interrupt": {"turn_id": str},
    "interrupt_resolved": {"turn_id": str},
    "ready": {
        "session_id": str,
        "balance_paise": int,
        "price_per_minute_paise": int,
        "tts_sample_rate_hz": int,
    },
    "response_cue": {
        "turn_id": str,
        "cue_id": str,
        "cue_key": str,
        "language_code": str,
        "delay_ms": int,
        "last_speech_capture_time_ms": (int, float),
    },
    "response_cue_cancel": {
        "turn_id": str,
        "cue_id": str,
        "reason": str,
    },
    "tool_request": {"call_id": str, "name": str, "arguments": dict},
    "turn_complete": {"turn_id": str},
    "turn_metrics": {"turn_id": str, "stages": dict},
}


def encode_capture_frame(frame: CaptureFrame) -> bytes:
    if frame.capture_seq < 0:
        raise ProtocolError("capture_seq must be non-negative")
    if not math.isfinite(frame.capture_time_ms):
        raise ProtocolError("capture_time_ms must be finite")
    if not frame.pcm or len(frame.pcm) % 2:
        raise ProtocolError("PCM payload must contain complete 16-bit samples")
    return (
        CAPTURE_HEADER.pack(
            CAPTURE_MAGIC,
            CAPTURE_VERSION,
            frame.capture_seq,
            frame.capture_time_ms,
        )
        + frame.pcm
    )


def decode_capture_frame(message: bytes) -> CaptureFrame:
    if len(message) <= CAPTURE_HEADER_BYTES:
        raise ProtocolError("binary frame is shorter than its header")
    magic, version, capture_seq, capture_time_ms = CAPTURE_HEADER.unpack_from(message)
    if magic != CAPTURE_MAGIC:
        raise ProtocolError("binary frame has an invalid magic value")
    if version != CAPTURE_VERSION:
        raise ProtocolError(f"unsupported capture frame version: {version}")
    if not math.isfinite(capture_time_ms):
        raise ProtocolError("capture_time_ms must be finite")
    pcm = message[CAPTURE_HEADER_BYTES:]
    if len(pcm) % 2:
        raise ProtocolError("PCM payload must contain complete 16-bit samples")
    return CaptureFrame(capture_seq=capture_seq, capture_time_ms=capture_time_ms, pcm=pcm)


def decode_tool_result(message: dict[str, Any]) -> ClientToolResult:
    call_id = message.get("call_id")
    if not isinstance(call_id, str) or not call_id:
        raise ProtocolError("tool_result requires a non-empty call_id")
    if "result" not in message:
        raise ProtocolError("tool_result requires a result field")
    return ClientToolResult(call_id=call_id, result=message["result"])


def validate_client_message(message: dict[str, Any]) -> dict[str, Any]:
    """Validate one client control message from the public WebSocket protocol."""

    validated = _validate_json_message(message, _CLIENT_REQUIRED_FIELDS, "client")
    message_type = validated["type"]
    if message_type == "interrupt_ack" and validated["audio_queue_cleared"] is not True:
        raise ProtocolError("interrupt_ack must confirm audio_queue_cleared")
    if message_type == "interrupt_ack":
        _require_order(validated, "interrupt_received_perf_ms", "queue_cleared_perf_ms")
    if message_type == "playback_started":
        ordered_fields = [
            "audio_received_perf_ms",
            "decode_complete_perf_ms",
            "scheduled_perf_ms",
            "playback_start_perf_ms",
        ]
        present = [field for field in ordered_fields if field in validated]
        for start, end in zip(present, present[1:], strict=False):
            _require_order(validated, start, end)
    if message_type == "response_cue_started" and (
        validated["cue_start_perf_ms"] < 0 or validated["feedback_voice_to_voice_ms"] < 0
    ):
        raise ProtocolError("response_cue_started timestamps must be non-negative")
    return validated


def validate_server_message(message: dict[str, Any]) -> dict[str, Any]:
    """Validate one server event from the public WebSocket protocol."""

    validated = _validate_json_message(message, _SERVER_REQUIRED_FIELDS, "server")
    message_type = validated["type"]
    if message_type == "agent_state" and validated["state"] not in {
        "interrupted",
        "listening",
        "speaking",
        "thinking",
        "user_speaking",
    }:
        raise ProtocolError("agent_state.state is invalid")
    if message_type == "call_ended" and validated["reason"] not in {
        "balance_exhausted",
        "error",
        "user",
    }:
        raise ProtocolError("call_ended.reason is invalid")
    if message_type == "response_cue":
        if validated["cue_key"] != "neutral_ack":
            raise ProtocolError("response_cue.cue_key is invalid")
        if validated["language_code"] not in {*CUE_LANGUAGES, "neutral"}:
            raise ProtocolError("response_cue.language_code is invalid")
        if validated["delay_ms"] < 0:
            raise ProtocolError("response_cue.delay_ms must be non-negative")
        if validated["last_speech_capture_time_ms"] < 0:
            raise ProtocolError("response_cue speech anchor must be non-negative")
    if message_type == "response_cue_cancel" and validated["reason"] not in {
        "answer_started",
        "call_ended",
        "new_user_speech",
        "turn_cancelled",
    }:
        raise ProtocolError("response_cue_cancel.reason is invalid")
    return validated


def _validate_json_message(
    message: dict[str, Any],
    schemas: dict[str, dict[str, type | tuple[type, ...]]],
    direction: str,
) -> dict[str, Any]:
    if not isinstance(message, dict):
        raise ProtocolError(f"{direction} message must be a JSON object")
    message_type = message.get("type")
    if not isinstance(message_type, str) or message_type not in schemas:
        raise ProtocolError(f"unknown {direction} message type: {message_type!r}")
    for field_name, expected_type in schemas[message_type].items():
        if field_name not in message:
            raise ProtocolError(f"{message_type} requires {field_name}")
        value = message[field_name]
        if expected_type is object:
            continue
        if expected_type in {(int, float), int} and isinstance(value, bool):
            raise ProtocolError(f"{message_type}.{field_name} has an invalid type")
        if not isinstance(value, expected_type):
            raise ProtocolError(f"{message_type}.{field_name} has an invalid type")
        if isinstance(value, str) and not value:
            raise ProtocolError(f"{message_type}.{field_name} cannot be empty")
        if isinstance(value, int | float) and not math.isfinite(value):
            raise ProtocolError(f"{message_type}.{field_name} must be finite")
    return dict(message)


def _require_order(message: dict[str, Any], start: str, end: str) -> None:
    if float(message[end]) < float(message[start]):
        raise ProtocolError(f"{start} must not follow {end}")


def sanitize_vendor_message(message: dict[str, Any]) -> dict[str, Any]:
    sanitized = _sanitize_value(message)
    if not isinstance(sanitized, dict):
        raise ProtocolError("vendor message must be a JSON object")
    return sanitized


def _sanitize_value(value: Any) -> Any:
    if isinstance(value, dict):
        result: dict[str, Any] = {}
        for key, item in value.items():
            if key in {"audio", "data"} and isinstance(item, str) and len(item) > 256:
                result[key] = f"<base64:{len(item)} chars>"
            else:
                result[key] = _sanitize_value(item)
        return result
    if isinstance(value, list):
        return [_sanitize_value(item) for item in value]
    return value

File: frontend/src/voice/protocol.ts
const CAPTURE_MAGIC = [0x41, 0x44, 0x53, 0x48] as const;
const CAPTURE_VERSION = 1;
const CAPTURE_HEADER_BYTES = 21;

export function encodeCaptureFrame(
  captureSeq: number,
  captureTimeMs: number,
  pcm: ArrayBuffer,
): ArrayBuffer {
  const message = new ArrayBuffer(CAPTURE_HEADER_BYTES + pcm.byteLength);
  const view = new DataView(message);
  CAPTURE_MAGIC.forEach((byte, index) => view.setUint8(index, byte));
  view.setUint8(4, CAPTURE_VERSION);
  view.setBigUint64(5, BigInt(captureSeq), false);
  view.setFloat64(13, captureTimeMs, false);
  new Uint8Array(message, CAPTURE_HEADER_BYTES).set(new Uint8Array(pcm));
  return message;
}

export type ServerEvent =
  | {
      type: "ready";
      session_id: string;
      balance_paise: number;
      price_per_minute_paise: number;
      tts_sample_rate_hz: number;
      endpointing_strategy: "local_vad" | "sarvam";
      vad_model?: string;
      capabilities?: string[];
    }
  | {
      type: "billing";
      session_id: string;
      seconds: number;
      cost_paise: number;
      session_cost_paise: number;
      charged_paise: number;
      balance_paise: number;
      low_balance: boolean;
      warning: "low_balance" | null;
      final: boolean;
      terminated_reason: "balance_exhausted" | null;
    }
  | {
      type: "agent_state";
      state: "listening" | "user_speaking" | "thinking" | "speaking" | "interrupted";
      transmitting?: boolean;
      transport_status?: "transmitting_speech" | "silence_not_transmitting";
      detail?: string;
    }
  | {
      type: "final_transcript";
      turn_id: string;
      text: string;
      language_code?: string | null;
      language_probability?: number | null;
    }
  | { type: "agent_text"; text: string; turn_id: string }
  | {
      type: "audio_start";
      turn_id: string;
      last_speech_capture_seq: number;
      last_speech_capture_time_ms: number;
      speech_anchor_source?: string;
    }
  | {
      type: "response_cue";
      turn_id: string;
      cue_id: string;
      cue_key: "neutral_ack";
      language_code: string;
      delay_ms: number;
      last_speech_capture_time_ms: number;
    }
  | {
      type: "response_cue_cancel";
      turn_id: string;
      cue_id: string;
      reason: "answer_started" | "call_ended" | "new_user_speech" | "turn_cancelled";
    }
  | { type: "turn_complete"; turn_id: string }
  | { type: "interrupt"; turn_id: string; reason: "barge_in" }
  | {
      type: "interrupt_resolved";
      turn_id: string;
      barge_in_stop_ack_ms: number | null;
      played_audio_ms: number;
      tts_socket_teardown_status: "closed_and_discarded" | "already_closed";
    }
  | {
      type: "tool_request";
      call_id: string;
      name: "todo_add" | "todo_list" | "todo_complete" | "todo_delete";
      arguments: Record<string, unknown>;
    }
  | TurnMetricsEvent
  | { type: "call_ended"; reason: string }
  | { type: "error"; code?: string; message: string; balance_paise?: number };

export type TurnMetricsEvent = {
  type: "turn_metrics";
  turn_id: string;
  endpointing_strategy: "local_vad" | "sarvam";
  stages: Record<string, number | null>;
  derived: Record<string, number | null>;
  dimensions: Record<string, unknown>;
  tool_spans: Array<{
    name: string;
    call_id: string;
    duration_ms: number;
    outcome: string;
  }>;
  missing_timestamps: string[];
};

export function parseServerEvent(message: string): ServerEvent {
  const parsed = JSON.parse(message) as unknown;
  if (typeof parsed !== "object" || parsed == null || !("type" in parsed)) {
    throw new Error("Invalid voice server event");
  }
  const event = parsed as Record<string, unknown>;
  if (event.type === "response_cue") {
    if (
      typeof event.turn_id !== "string" ||
      typeof event.cue_id !== "string" ||
      event.cue_key !== "neutral_ack" ||
      typeof event.language_code !== "string" ||
      typeof event.delay_ms !== "number" ||
      !Number.isFinite(event.delay_ms) ||
      event.delay_ms < 0 ||
      typeof event.last_speech_capture_time_ms !== "number" ||
      !Number.isFinite(event.last_speech_capture_time_ms)
    ) {
      throw new Error("Invalid response cue event");
    }
  }
  if (
    event.type === "response_cue_cancel" &&
    (typeof event.turn_id !== "string" || typeof event.cue_id !== "string")
  ) {
    throw new Error("Invalid response cue cancellation");
  }
  return event as ServerEvent;
}
